docs/source-2.0/guides/codegen/mapping-shapes-to-languages.rst

```diff
@@ -0,0 +1,629 @@
+traits. You can transform the model being code generated and create
+synthetic input and output shapes when necessary using the
+`createDedicatedInputAndOutput <https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/transform/ModelTransformer.java#L546>`_
+model transformer. The following example creates a new Model that has
+dedicated input and output shapes for every operation, each marked with
+the ``@input`` or ``@output`` trait, and each uses a consistent name
+that ends with ``Input`` or ``Output``.
+
+.. code-block:: java
+
+   ModelTransformer transformer = ModelTransformer.create();
+   Model transformed = transformer.createDedicatedInputAndOutput(
+       "model", "Input", "Output"
+   );
+
+.. tip::
+
+    :ref:`directedcodegen` automatically handles this for you.
+
+
+Resource
+--------
+
+A *resource* is an entity with an identity that has a set of operations.
+
+Resources add hierarchy to a model. You will need to traverse from a
+service to every operation and resource in order to crawl the entire
+service. This process can be simplified using the
+`TopDownIndex <https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TopDownIndex.java>`__.
+Iterating only over the operations attached to a service will not
+provide every operation in the closure of the service.
+
+.. note::
+
+    Exposing resource abstractions through code, if attempted, should
+    be done in addition to a more traditional, flattened, service
+    interface with every operation contained in the service.
+
+
+Other shape topics
+==================
+
+Shapes can be recursive
+-----------------------
+
+Smithy :ref:`shapes support recursion <recursive-shape-definitions>`.
+Some languages like Rust require the size of types to be known at
+compile time. Recursive types in these languages need some kind of heap
+allocation to ensure they have a size known at compile time. In order to
+identify which member in a recursive loop needs to be heap allocated,
+implementations will need to utilize a topological sort. Smithy's
+:ref:`directedcodegen` abstraction will automatically generate code based
+on a topological sort, though generators that need more control over how to handle recursion will need to manually
+use a `TopologicalIndex <https://github.com/awslabs/smithy/blob/main/smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TopologicalIndex.java>`__.
+
+.. seealso::
+
+        `Rust design doc <https://github.com/awslabs/smithy-rs/blob/main/design/src/smithy/recursive_shapes.md>`__
+        for how they handled recursive shapes.
+
+
+Mixins are an implementation detail of the model
+------------------------------------------------
+
+Mixins are considered an implementation detail of a model and should not
+impact code generation. Code generators should transform the model prior
+to code generation to remove mixins.
```

Comment:
Call out the transformer method like for input/output?

Reply:
Flattening them you mean? I give a link below in the seealso directive.

Reply:
Two docs blurred while reading this, all good on current link.